src/imperative/imperative_utils.h

```diff
@@ -35,6 +35,59 @@
 #define MXNET_IMPERATIVE_IMPERATIVE_UTILS_H_
 
 namespace mxnet {
+
+#if MXNET_USE_MKLDNN == 1
+template<typename T>T *pntr(T &obj)           { return &obj; }  // NOLINT
+template<typename T>T *pntr(T *obj)           { return obj; }
+
+template<typename T>
+void InvalidateOutputs(const std::vector<T> *pArrs,
+                        const std::vector<OpReqType> &reqs) {
+  auto arrs = *pArrs;
+  for (size_t i = 0; i < arrs.size(); i++) {
+    if (reqs[i] == kWriteTo || reqs[i] == kNullOp)
+      pntr(arrs[i])->InvalidateMKLDNNData();
+  }
+}
+
+static inline void CreateDefaultInputs(const std::vector<NDArray> &arrs,
+                                       std::vector<NDArray> *out_arrs) {
+  out_arrs->clear();
+  for (size_t i = 0; i < arrs.size(); ++i) {
+    if (arrs[i].IsMKLDNNData())
+      out_arrs->push_back(arrs[i].Reorder2Default());
+    else
+      out_arrs->push_back(arrs[i]);
+  }
+}
+
+static inline void CreateDefaultInputs(std::vector<NDArray> *pArrs) {
+  auto &&arrs = *pArrs;
+  for (size_t i = 0; i < arrs.size(); ++i)
+    arrs[i].SelfReorder2Default();
+}
+
+#define INVALIDATE_OUTPUTS(outputs, req) InvalidateOutputs(&outputs, req)
```

Comment:
Why do you make this macros instead of calling the function directly?

Reply:
As I mentioned in the PR **Description** (see **Changes**), that and a couple of other macros are defined differently for different values of `MXNET_USE_MKLDNN` flag:  
```
#if MXNET_USE_MKLDNN == 1
#define INVALIDATE_OUTPUTS(outputs, req) InvalidateOutputs(&outputs, req)
#else
#define INVALIDATE_OUTPUTS(outputs, ...)       // empty macros
#endif
```

It allows us to eliminate the multiple usages of `#if MXNET_USE_MKLDNN == 1 / #endif lines and simplify the code.

Reply:
Thank you for simplifying the code. Do you think it would be better to rely on `constexpr if` inside the functions (`IsDefaultData`, `InvalidateOutputs`, ...) instead of adding more Macros to the codebase?

Reply:
Frankly, I am not familiar with  `constexpr if` notion. I will take a look at it, to see if it is possible to use it as effective as empty macros.

Reply:
> Do you think it would be better to rely on `constexpr if` inside the functions (`IsDefaultData`, `InvalidateOutputs`, ...) instead of adding more Macros to the codebase?

Definitely, it could be done, but I am not sure that it will be as effective as the use of empty macros. My guess is: the compiler is not smart enough to DO NOT generate the functions `IsDefaultData, InvalidateOutputs, ...` when they are empty.  Therefore, they will be defined, called, and do nothing! 

I think that the use of the trick with empty macro is more effective: 

-  we define, compile, and use functions only when they really needed. 
-  the types of parameters of such functions could be defined ONLY when they will be really used.  (Therefore, in general, we don't need even the presence of some header files with these definitions)

But I could implement these functions with  `constexpr if` if you really think that it would be better.

Reply:
> My guess is: the compiler is not smart enough to DO NOT generate the functions IsDefaultData, InvalidateOutputs, ... when they are empty. Therefore, they will be defined, called, and do nothing!

My guess was that such small (empty) functions would be inlined. But it's fine to keep the empty macro trick